### PR TITLE
[`pyproject.toml`: part 3] Add means to apply configuration from `pyproject.toml`

### PR DESCRIPTION
**Pull request by abravalheri**
> *This is part of #2970, split into several stacked PRs as requested in https://github.com/pypa/setuptools/pull/2970#pullrequestreview-868151658*
> *Follows #3066*

This change targets adding a standardised project configuration format as initially proposed by PEP 621 (with the possibility of modification/evolution according to follow-up PEPs).

For the configuration fields not covered by the standards, the approach proposed in [`ini2toml`](https://ini2toml.readthedocs.io/en/latest/setuptools_pep621.html) is adopted.

This operation relies on vendorised external libraries `tomli` and `validate-pyproject`.

## Summary of changes

- Implement `read_configuration` from `pyproject.toml` - 3eb62ac2912f8c0a8bfbdc5dbff809f983a14e2a
- Expand dynamic `entry_points` from `pyproject.toml` - 112e8a930314eddf9f35159ea890aaec750843c8
- Add means of applying config read from `pyproject.toml` to dist - 303791af3ed6be54d0a3663b1738537be6b30f32
- Add the `apply_configuration` API to `setuptools.config.setupcfg` - 43e34361e072880841654412c61d9c508e28a5f9
- Test `pyproject.toml` config has the same effect as `setup.cfg` -  c6907cb008fd647f9180e29ec4ee780e148b832c

Please see the commit message of each individual commit for more information about the changes.

### Pull Request Checklist
- [x] Changes have tests
- [ ] News fragment added in [`changelog.d/`].
  _(See [documentation][PR docs] for details)_

[`changelog.d/`]: https://github.com/pypa/setuptools/tree/master/changelog.d
[PR docs]:
https://setuptools.pypa.io/en/latest/development/developer-guide.html#making-a-pull-request

### CODE REVIEW

**Comment by webknjaz** [comment]
@abravalheri FYI the PR title has a typo (missing "t" in "pyproject"), I can't fix it myself.

File: setuptools/tests/config/test_apply_pyprojecttoml.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+
+EXAMPLES = (Path(__file__).parent / "setupcfg_examples.txt").read_text()
+EXAMPLE_URLS = [x for x in EXAMPLES.splitlines() if not x.startswith("#")]
+DOWNLOAD_DIR = Path(__file__).parent / "downloads"
+
+
+@pytest.mark.parametrize("url", EXAMPLE_URLS)
+@pytest.mark.filterwarnings("ignore")
+def test_apply_pyproject_equivalent_to_setupcfg(url, monkeypatch, tmp_path):
```
Comment (webknjaz) [inline setuptools/tests/config/test_apply_pyprojecttoml.py:26]: Am I reading this right that this relies on network I/O? Maybe this could be marked as such?

File: setuptools/tests/config/test_apply_pyprojecttoml.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        file_name = file_name.replace(part, '').strip().strip('/:').strip()
+    file_name = re.sub(r"[^\-_\.\w\d]+", "_", file_name)
+    path = Path(download_dir, file_name)
+    if not path.exists():
+        download_dir.mkdir(exist_ok=True, parents=True)
+        download(url, path)
+    return path
+
+
+def download(url, dest):
```
Comment (webknjaz) [inline setuptools/tests/config/test_apply_pyprojecttoml.py:161]: It's probably worth implementing a global cache and avoiding making fixture download part of testing. It should be possible to request cache download or `--offline` via a pytest CLI flag that could be implemented via a pytest hook callback.

File: setuptools/tests/config/test_apply_pyprojecttoml.py (hunk 1 above)
Comment (abravalheri) [reply setuptools/tests/config/test_apply_pyprojecttoml.py:26]: Yes @webknjaz, this test will try to download a bunch of real world examples... It is possible to cache in the CI, but I haven't implemented that yet.

Is your suggestion adding a simple `@pytest.mark.uses_network`, or something more complex?

File: .gitignore [hunk 3]
```diff
@@ -20,3 +20,4 @@ setuptools.egg-info
 .idea/
 .pytest_cache/
 .mypy_cache/
+/setuptools/tests/config/downloads
```
Comment (webknjaz) [inline .gitignore:23]: Instead of this, I suggest creating `setuptools/tests/config/downloads/.gitignore` that contains:
```gitignore
*
!.gitignore
```

File: setuptools/tests/config/test_apply_pyprojecttoml.py (hunk 1 above)
Comment (abravalheri) [reply setuptools/tests/config/test_apply_pyprojecttoml.py:26]: (when running the tests locally, the files will be cached)

File: setuptools/tests/config/test_apply_pyprojecttoml.py (hunk 2 above)
Comment (webknjaz) [reply setuptools/tests/config/test_apply_pyprojecttoml.py:161]: Some of related ideas have been suggested in the past @ https://github.com/pypa/setuptools/discussions/2607

File: setuptools/tests/config/test_apply_pyprojecttoml.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    value = "\n".join(buffer.getvalue().strip().splitlines())
+
+    # ---- DIFF NORMALISATION ----
+    # PEP 621 is very particular about author/maintainer metadata conversion, so skip
+    value = re.sub(r"^(Author|Maintainer)(-email)?:.*$", "", value, flags=re.M)
+    # May be redundant with Home-page
+    value = re.sub(r"^Project-URL: Homepage,.*$", "", value, flags=re.M)
+    # May be missing in original (relying on default) but backfilled in the TOML
+    value = re.sub(r"^Description-Content-Type:.*$", "", value, flags=re.M)
+    # ini2toml can automatically convert `tests_require` to `testing` extra
```
Comment (webknjaz) [inline setuptools/tests/config/test_apply_pyprojecttoml.py:183]: Why not "tests"? I would expect that "testing" contains helpers for the end-users to test their project using the given package while "tests" would correspond to its own internal test requirements.

File: setuptools/tests/config/test_pyprojecttoml.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+    tool = {"setuptools": {"dynamic": {"entry-points": {"file": "entry-points.txt"}}}}
+    project = {"dynamic": ["scripts", "gui-scripts", "entry-points"]}
+    pyproject = {"project": project, "tool": tool}
+    expanded = expand_configuration(pyproject, tmp_path)
+    assert len(expanded["project"]["scripts"]) == 1
+    assert expanded["project"]["scripts"]["a"] == "mod.a:func"
+    assert len(expanded["project"]["gui-scripts"]) == 1
+    assert expanded["project"]["gui-scripts"]["b"] == "mod.b:func"
+    print(expanded["project"]["entry-points"])
```
Comment (webknjaz) [inline setuptools/tests/config/test_pyprojecttoml.py:127]: Why is this necessary? I'm pretty sure pytest can show the var contents on failures...

File: setuptools/tests/config/test_pyprojecttoml.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    entry_points = ConfigParser()
+    entry_points.read_dict(ENTRY_POINTS)
+    with open(tmp_path / "entry-points.txt", "w") as f:
+        entry_points.write(f)
+
+    tool = {"setuptools": {"dynamic": {"entry-points": {"file": "entry-points.txt"}}}}
+    project = {"dynamic": ["scripts", "gui-scripts", "entry-points"]}
+    pyproject = {"project": project, "tool": tool}
+    expanded = expand_configuration(pyproject, tmp_path)
+    assert len(expanded["project"]["scripts"]) == 1
```
Comment (webknjaz) [inline setuptools/tests/config/test_pyprojecttoml.py:123]: It'd be nice to put `expanded["project"]` in a var to improve readability.

File: setuptools/tests/config/test_apply_pyprojecttoml.py (hunk 4 above)
Comment (abravalheri) [reply setuptools/tests/config/test_apply_pyprojecttoml.py:183]: I ended up going with `testing` because this seems to be popular in the community... I have the impression that this approach was influenced by the fact that [tox docs](https://tox.wiki/en/latest/config.html?highlight=extras#conf-extras) explicitly mention `testing`.

For example [setuptools](https://github.com/pypa/setuptools/blob/fe7975d091483e8d6d4c9befea38ccfeb29e8e49/setup.cfg#L41) and [virtualenv](https://github.com/pypa/virtualenv/blob/7424a015d1c28e26fd9324b8c47dab6d6c15b184/setup.cfg#L95) use `testing`.

File: setuptools/tests/config/test_pyprojecttoml.py (hunk 5 above)
Comment (abravalheri) [reply setuptools/tests/config/test_pyprojecttoml.py]: Probably something left behind a debug session (I think `pytest` displays only a limited amount of chars by default).

Thank you very much for pointing it out!

**Comment by abravalheri** [comment]
Thank you very much for the review @webknjaz.

Regarding the offline tests, I tried to mark the tests here so they can be skipped. I am afraid implementing `--offline` in pytest will be a much bigger scope and involve changes in existing tests, so I am deferring that for the future.

File: setuptools/tests/config/setupcfg_examples.txt
Comment (webknjaz) [inline setuptools/tests/config/setupcfg_examples.txt:1]: @abravalheri so I rediscovered this file by accident and wanted to say that I saw a corner case with `setup.cfg` behaving differently from `pyproject.toml` in the case of multiple package roots (in that `pyproject.toml` works while `setup.cfg` doesn't): https://github.com/ansible/ansible/pull/81440/files#diff-fa602a8a75dc9dcc92261bac5f533c2a85e34fcceaff63b3a3a81d9acde2fc52R40-R42.
